src/libraries/icubmod/bcbBattery/bcbBattery.h

```diff
@@ -8,61 +8,83 @@
 #define __BCBBATTERY_H__
 
 #include <mutex>
-#include <atomic>
 #include <yarp/os/PeriodicThread.h>
 #include <yarp/dev/IBattery.h>
 #include <yarp/dev/PolyDriver.h>
 #include <yarp/dev/SerialInterfaces.h>
 #include <yarp/os/ResourceFinder.h>
 #include <yarp/sig/Vector.h>
+#include <regex>
 
 using namespace yarp::os;
 using namespace yarp::dev;
 
-class BcbBattery : public PeriodicThread, public yarp::dev::IBattery, public DeviceDriver
+class batteryReaderThread : public PeriodicThread
 {
-protected:
-    std::mutex mtx;
+    public:
+    //configuration options
+    bool               verboseEnable = false;
+    bool               screenEnable = true;
+    bool               silenceSyncWarnings = false;
 
+    //the buffer
     double             timeStamp;
-    double             battery_charge;
-    double             battery_voltage;
-    double             battery_current;
-    std::string        battery_info;
-    unsigned char      backpack_status;
+    static const int   buff_len = 10000;
+    unsigned char      tmp_buff[buff_len];
+    static const int   packet_len =10;
+    char               packet[packet_len];
+    std::regex         r_exp;
+
+    ISerialDevice*     pSerial = nullptr;
+    double             battery_charge = 0;
+    double             battery_voltage = 0;
+    double             battery_current = 0;
+    std::string        battery_info = "icub battery system v1.0";
+    int                backpack_status = 0;
+    IBattery::Battery_status     battery_status = IBattery::Battery_status::BATTERY_OK_STANBY;
+
+    batteryReaderThread (ISerialDevice *_pSerial, double period) :
+    PeriodicThread((double)period),
+    pSerial(_pSerial)
+    {
+       //                         0   1   2   3   4   5   6   7   8    9
+       char c_exp[packet_len +5+ packet_len +1] =
+                               {'\0','.','.','.','.','.','.','.','\r','\n',
+                                 '(','?','!','.','*',
+                                '\0','.','.','.','.','.','.','.','\r','\n',
+                                 ')'};
```

Comment:
I would suggest adding a comment with an explanation of the expressions to ease future modifications.

Reply:
done